File: 卡码/代码随想录/Python/二叉树/02-二叉树属性/LeetCode_0572.py
#572. 另一棵树的子树
from typing import *
from ..TreeNode import *

class Solution:
  def compare(self,left:Optional[TreeNode],right:Optional[TreeNode])->bool:
    if left!=None and right==None or left==None and right!=None:
      return False
    elif left==None and right==None:
      return True
    elif left.val!=right.val:
      return False
    else:
      return self.compare(left.left,right.left) and self.compare(left.right,right.right)

  def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
    if not root:
      return False
    return self.compare(root,subRoot) or self.isSubtree(root.left,subRoot) or self.isSubtree(root.right,subRoot)

File: 卡码/代码随想录/JS/二叉树/02-二叉树属性/LeetCode_0572.js
//572. 另一棵树的子树
import TreeNode from "../TreeNode.js";

const compare=(left,right)=>{
  if(left&&!right) return false;
  else if(!left&&right) return false;
  else if(!left&&!right) return true;
  else if(left.val!=right.val) return false;
  else return compare(left.left,right.left)&&compare(left.right,right.right);
};

const isSubtree = function(root, subRoot) {
  if(!root) return false;
  return compare(root,subRoot)||isSubtree(root.left,subRoot)||isSubtree(root.right,subRoot);
};